[Optim in backward] API to retrieve in-backward optimizers

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105995
* __->__ #105991

API to retrieve in backward optimizer for checkpointing purposes

Differential Revision: [D47782225](https://our.internmc.facebook.com/intern/diff/D47782225/)

rohan-varma committed:
[Optim in backward] API to retrieve in-backward optimizers

API to retrieve in backward optimizer for checkpointing purposes

Differential Revision: [D47782225](https://our.internmc.facebook.com/intern/diff/D47782225/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Optim in backward] API to retrieve in-backward optimizers"


API to retrieve in backward optimizer for checkpointing purposes

Differential Revision: [D47782225](https://our.internmc.facebook.com/intern/diff/D47782225/)

[ghstack-poisoned]

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -88,3 +88,26 @@ def optimizer_hook(*_unused) -> None:
         _apply_optimizer_in_backward_to_param(param)
+
+
+def _get_optimizers_in_backward(model: torch.nn.Module) -> List[torch.optim.Optimizer]:
```
Do you think `_get_in_backward_optimizers()` would read better? Does TorchRec have an API like this; if so, what is it called?

awgu commented on test/distributed/optim/test_apply_optimizer_in_backward.py:
```diff
@@ -137,3 +137,19 @@ def test_multiple_optim_for_params(self) -> None:
+
+        # Apply optimizers in backward
+        _apply_optimizer_in_backward(torch.optim.SGD, model.parameters(), {'lr': 0.01})
+        result = set(_get_optimizers_in_backward(model))
+        expected = set([optim for p in model.parameters() for optim in p._in_backward_optimizers])
```
If the new API returns a list, should we be testing the order too?

rohan-varma replied:
I don't think the order matters too much - if it does, we should probably be returning something like a map of FQN to the optimizer. @fegin , do you expect the order to matter for checkpointing?

and if order indeed does not matter, is there an alternative return type than list?

fegin replied:
The order does not matter. If users want to know what is optimized by which optimizer, users can just use `optimzer.param_groups`.

fegin commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -88,3 +88,26 @@ def optimizer_hook(*_unused) -> None:
+    """
+    optims = []
+    for param in model.parameters():
+        optims.extend(param._in_backward_optimizers)
```
Should we use `getattr` and also check if `_in_backward_optimizers` exists in the param?

rohan-varma committed:
Update on "[Optim in backward] API to retrieve in-backward optimizers"


API to retrieve in backward optimizer for checkpointing purposes

Differential Revision: [D47782225](https://our.internmc.facebook.com/intern/diff/D47782225/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Optim in backward] API to retrieve in-backward optimizers"


API to retrieve in backward optimizer for checkpointing purposes

Differential Revision: [D47782225](https://our.internmc.facebook.com/intern/diff/D47782225/)

[ghstack-poisoned]

awgu commented on test/distributed/optim/test_apply_optimizer_in_backward.py:
```diff
@@ -137,3 +140,21 @@ def test_multiple_optim_for_params(self) -> None:
+
+        # Apply optimizers in backward
+        _apply_optimizer_in_backward(torch.optim.SGD, model.parameters(), {"lr": 0.01})
+        result = set(_get_in_backward_optimizers(model))
```
How are optimizers hashed? (I am guessing just by object.)

nit: Maybe you should check that `len(_get_in_backward_optimizers(model))` is expected (i.e. equal to `len(list(model.parameters()))`)?

awgu commented on torch/distributed/optim/apply_optimizer_in_backward.py:
```diff
@@ -88,3 +88,26 @@ def optimizer_hook(*_unused) -> None:
         _apply_optimizer_in_backward_to_param(param)
+
+
+def _get_in_backward_optimizers(model: torch.nn.Module) -> List[torch.optim.Optimizer]:
```
Just as a general comment: If we do not expect a function to be strictly called on the root module, I think a good convention is to make the arg name `module` (across all our APIs) instead of `model`. (For us, I think we mainly consider this in terms of composing with a sparse component or with other distributed APIs.)

rohan-varma committed:
Update on "[Optim in backward] API to retrieve in-backward optimizers"


API to retrieve in backward optimizer for checkpointing purposes

Differential Revision: [D47782225](https://our.internmc.facebook.com/intern/diff/D47782225/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Optim in backward] API to retrieve in-backward optimizers"


API to retrieve in backward optimizer for checkpointing purposes

Differential Revision: [D47782225](https://our.internmc.facebook.com/intern/diff/D47782225/)

[ghstack-poisoned]
